src/transformers/models/xy_tokenizer/modeling_xy_tokenizer.py

```diff
@@ -0,0 +1,1692 @@
+        cluster_size = (
+            (self.cluster_size + self.epsilon)
+            / (self.cluster_size.sum() + self.codebook_size * self.epsilon)
+            * self.cluster_size.sum()
+        )
+        self.codebook.copy_(self.embed_avg / cluster_size.unsqueeze(1))
+
+    def replace_dead_codes(self, encodings):
+        if self.threshold_ema_dead == 0:
+            return
+        dead_mask = self.cluster_size < self.threshold_ema_dead
+        if dead_mask.any():
+            samples = (
+                sample_vectors(encodings.float(), self.codebook_size)
+                if not dist.is_initialized() or dist.get_rank() == 0
+                else torch.zeros_like(self.codebook)
+            )
+            if dist.is_initialized():
+                dist.broadcast(samples, src=0)
+            self.codebook[dead_mask] = samples[: dead_mask.sum()].to(self.codebook.dtype)
+
+    def init_codebook(self, encodings):
+        if self.inited.item():
+            return
+        if not dist.is_initialized() or dist.get_rank() == 0:
+            embed, cluster_sizes = kmeans(encodings.float(), self.codebook_size, self.kmeans_iters)
+        else:
+            embed, cluster_sizes = (
+                torch.zeros(self.codebook_size, self.codebook_dim, device=encodings.device),
+                torch.zeros(self.codebook_size, device=encodings.device),
+            )
+        if dist.is_initialized():
+            dist.broadcast(embed, src=0)
+            dist.broadcast(cluster_sizes, src=0)
+        self.codebook.copy_(embed)
+        self.embed_avg.copy_(embed.clone())
+        self.cluster_size.copy_(cluster_sizes)
+        self.inited.fill_(True)
+
+    def forward(self, z):
+        z_e = self.in_project(z.float())
+        encodings = z_e.permute(0, 2, 1).reshape(-1, z_e.size(1))  # b d t -> (b t) d
+        if self.kmeans_init and not self.inited.item():
+            self.init_codebook(encodings)
+        dist = (
+            encodings.pow(2).sum(1, keepdim=True)
+            - 2 * encodings @ self.codebook.float().t()
+            + self.codebook.float().pow(2).sum(1, keepdim=True).t()
+        )
+        indices = (-dist).max(1)[1].reshape(z.size(0), -1)  # (b t) -> b t
+        z_q_emb = F.embedding(indices, self.codebook.float()).transpose(1, 2)  # Get embeddings without projection
+        commit_loss = F.mse_loss(z_e, z_q_emb.detach(), reduction="none").mean([1, 2]) * self.commitment
+        if self.training and torch.is_grad_enabled():
+            self.ema_update(encodings, F.one_hot(indices.view(-1), self.codebook_size))
+            self.replace_dead_codes(encodings)
+        z_q = self.out_project(z_e + (z_q_emb - z_e).detach())
+        return z_q, commit_loss, torch.tensor(0.0, device=z.device), indices, z_e
+
+    def decode_code(self, embed_id):
+        embeddings = F.embedding(embed_id, self.codebook.float()).transpose(1, 2)
+        return self.out_project(embeddings)
+
+
+class ResidualVQ(nn.Module):
```

Comment:
Are there similarities with [DAC's RVQ](https://github.com/huggingface/transformers/blob/37c14430c99edca79dfcdcb76f1209f291b12fab/src/transformers/models/dac/modeling_dac.py#L103)? If so, we can simularly use modular to inherit/import from it.